src/sqlfluff/rules/L003.py

```diff
@@ -1,15 +1,156 @@
+        for field_el in dataclasses.fields(self):
+            key = field_el.name
+            yield key, getattr(self, key)
+
+
+@dataclasses.dataclass
+class _LineSummary(_DictLikeDataCls):
+    """A dataobject to represent a line.
+
+    Doubles as a memory object (holding state)
+    that can be converted to a fresh summary instance
+    """
+
+    line_no: int = 0
+    templated_line: Optional[int] = None
+    line_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    ignoreable_reference: List[bool] = dataclasses.field(default_factory=list)
+    indent_buffer: List[BaseSegment] = dataclasses.field(default_factory=list)
+    indent_size: int = 1
+    indent_balance: int = 0
+    hanging_indent: Optional[int] = None
+    clean_indent: bool = True
+    template_content: str = ""
+    templated_line_type: Optional[str] = None
+    line_anchor: Optional[BaseSegment] = None
+
+    # These are only required for carrying state between lines
+    anchor_indent_balance: int = 0
+    in_indent: bool = True
+    line_indent_stack: List[int] = dataclasses.field(default_factory=list)
+    hanger_pos: Optional[int] = None
+
+    def strip_buffers(self) -> dict:
+        """Strip a line dict of buffers for logging."""
+        keys_to_strip = ("line_buffer", "indent_buffer", "ignoreable_reference")
+        print_dict: Dict = {
+            key: value for key, value in self.items() if key not in keys_to_strip
+        }
+        print_dict["raw"] = "".join(el.raw for el in self.line_buffer)
+        return print_dict
+
+    def trigger_el(self):
+        """Get the pivot element."""
+        for el, is_ignorable in zip(self.line_buffer, self.ignoreable_reference):
+            if is_ignorable:
+                continue
+            if el.is_type("newline", "whitespace"):
+                continue
+            if el.segments:
+                continue
+            if el.is_meta and el.indent_val != 0:
+                continue
+            return el
+
+        return None
+
+    def memo_line_reset(self):
+        """When acting like a MemoryLine this resets state."""
+        self.templated_line = False
+        self.indent_buffer = []
+        line_buffer = self.line_buffer
+        self.line_buffer = []
+        self.line_indent_stack = []
+        self.ignoreable_reference = []
+        self.indent_size = 0
+        self.in_indent = True
+        self.hanger_pos = None
+        self.line_anchor = None
+        self.clean_indent = False
+        # Assume an unclean indent, but if the last line
+        # ended with an indent then we might be ok.
+        # Was there an indent after the last code element of the previous line?
+        for search_elem in reversed(line_buffer):
+            is_meta = search_elem.is_meta
+            if not search_elem.is_code and not is_meta:
+                continue
+            elif is_meta and search_elem.indent_val > 0:
+                self.clean_indent = True
+            break
```

Comment:
Could it be worth creating a separate function for this? (More self documenting.)